# Lab 3, part I: sensitivity analysis

This lab focuses on sensitivity analysis. After solving a problem with three constraints and four variables, we will change the problem slightly and re-solve it to evaluate how the solution and the objective function change. Then we'll use the problem's dual variables and reduced costs to arrive at the same result.

Consider the product mix problem: A factory produces 4 types of perfume by mixing 5 ingredients. Denote as $P=\{1,2,3,4\}$ the set of perfumes, and as $I=\{1,2,3,4,5\}$ the set of ingredients. The retail price for each liter of perfume $p\in P$ is $c_p$, defined in the table below:

|Perfume | 1 | 2 | 3 | 4 |
|--------|---|---|---|---|
|Retail price  |300|255|260|390|

One liter of perfume $p$ requires a quantity $a_{ip}$ of the ingredient $i$, also expressed in liters, as specified below:

|Ingredient\Perfume|1|2|3|4|
|-|-|-|-|-|
|1|0.01|0.05|0.07|0.04|
|2|0.34|0.45|0.36|0.51|
|3|0.08|0.06|0.12|0.12|
|4|0.55|0.35|0.29|0.32|
|5|0.02|0.09|0.16|0.01|

There is a finite amount (in liters) of each ingredient expressed by $b_i$, for all $i\in I$ and defined in the following table:

|Ingredient|1|2|3|4|5|
|----------|-|-|-|-|-|
|Availability|30|400|90|450|70|

Here are the tasks for this exercise:

1. Determine the amount of perfume to produce for each type in order to maximize the total revenue;
2. What could allow us to increase the total profit without changing the retail price $c_p$ and the percentage $a_{ip}$?
3. Find what constraint(s) are the most restrictive by changing one of the input parameters $b_i$;
4. Are all perfumes produced? If any of them is not produced, can we increase its retail price so that it is worth to produce it? By how much?

In [ ]:
# When using Colab, make sure you run this cell beforehand
!pip install mip

## Solution

Task 1: __Determine the amount of perfume to produce for each type in order to maximize the total revenue.__

We must create an optimization problem for this purpose. First, define sets and parameters:

* Set $P = \{1,2,3,4\}$;
* Set $I = \{1,2,3,4,5\}$;
* Retail price $c = (300,255,260,390)^\top$;
* Availability $b = (30,400,90,450,70)^\top$;
* Composition $A = \left(
\begin{array}{rrrrr}
0.01 & 0.05 & 0.07 & 0.04\\
0.34 & 0.45 & 0.36 & 0.51\\
0.08 & 0.06 & 0.12 & 0.12\\
0.55 & 0.35 & 0.29 & 0.32\\
0.02 & 0.09 & 0.16 & 0.01\\
\end{array}
\right)$

There is only one class of variables, $x_i$ for each $i\in P$, i.e. one nonnegative variable for each type of perfume. The problem is then as follows:

$$
\begin{array}{lrrrrrr}
\max          & 300 x_1 &+  255 x_2 &+  260 x_3  &+ 390 x_4\\
\textrm{s.t.} &0.01 x_1 &+ 0.05 x_2 &+ 0.07 x_3 &+ 0.04 x_4 &\le & 30\\
              &0.34 x_1 &+ 0.45 x_2 &+ 0.36 x_3 &+ 0.51 x_4 &\le &400\\
              &0.08 x_1 &+ 0.06 x_2 &+ 0.12 x_3 &+ 0.12 x_4 &\le & 90\\
              &0.55 x_1 &+ 0.35 x_2 &+ 0.29 x_3 &+ 0.32 x_4 &\le &450\\
              &0.02 x_1 &+ 0.09 x_2 &+ 0.16 x_3 &+ 0.01 x_4 &\le & 70\\
              &x_1, &x_2, &x_3, &x_4 &\ge& 0,
\end{array}
$$

or, more concisely,

$$
\begin{array}{ll}
\max& c^\top x\\
\textrm{s.t.} & Ax \le b\\
              & x \ge 0.
\end{array}
$$

We implement it below using Python-MIP.

Because we will solve this problem more than once, we create a function that takes $A$, $b$, and $c$ as arguments and returns the optimal solution found, its objective function value, and the model itself.

In [1]:
import mip

def solve_productmix(A, b, c):
    
    """
    Solve the problem max{cx: Ax <= b, x >= 0} with A, b, c passed as arguments
    """

    global P, I

    n = len(c)
    k = len(b)

    # create model (TODO)
    m = mip.Model()
    # add variables (TODO)
    x = [m.add_var() for j in range(n)]

    # One constraint per ingredient (TODO)
    for i in range(k):
        m+=mip.xsum(A[i][j]*x[j] for j in range(n))<=b[i]
    
    # Objective function is a weighted sum of all x (TODO)
    m.objective = mip.maximize(mip.xsum(c[j]*x[j] for j in range(n)))
    # Solve
    m.optimize()

    # Return a tuple containing model, solution, and objective value
    return (m, [x[p].x for p in P], m.objective_value)


c = [300, 255, 260, 390]
b = [30, 400, 90, 450, 70]

A = \
[[0.01, 0.05, 0.07, 0.04],
 [0.34, 0.45, 0.36, 0.51],
 [0.08, 0.06, 0.12, 0.12],
 [0.55, 0.35, 0.29, 0.32],
 [0.02, 0.09, 0.16, 0.01]]

nperf = len(c)  # number of perfumes is the length of c
ning = len(b)   # number of ingredients is the length of b

P = range(nperf)
I = range(ning)

# Use "_" because we don't care about the model in this call
_, solution, objective = solve_productmix(A,b,c)

print("Solution:", solution)
print("Objective: {0:10.2f}".format(objective))

Welcome to the CBC MILP Solver 
Version: Trunk
Build Date: Oct 28 2021 

Starting solution of the Linear programming problem using Primal Simplex

Coin0506I Presolve 5 (0) rows, 4 (0) columns and 20 (0) elements
Clp1000I sum of infeasibilities 0 - average 0, 1 fixed columns
Coin0506I Presolve 5 (0) rows, 3 (-1) columns and 15 (-5) elements
Clp0029I End of values pass after 3 iterations
Clp0000I Optimal - objective value 320809.22
Clp0000I Optimal - objective value 320809.22
Coin0511I After Postsolve, objective 320809.22, infeasibilities - dual 0 (0), primal 0 (0)
Clp0000I Optimal - objective value 320809.22
Clp0000I Optimal - objective value 320809.22
Clp0000I Optimal - objective value 320809.22
Clp0032I Optimal objective 320809.2155 - 0 iterations time 0.002, Idiot 0.00
Solution: [573.115003808073, 89.74358974359002, 0.0, 323.0515359228229]
Objective:  320809.22


Task 2: __What could allow us to increase the total profit without changing the retail price $c_p$ and the composition $a_{ip}$?__

If the retail price and the composition cannot change, we can only modify the availability. For example, by increasing all values of $b$ by 10%, we can easily increase the revenue by the same percentage:

In [2]:
# Create b2 that is 1.1 times b (TODO)
b2 = [1.1*i for i in b]

_, solution2, objective2 = solve_productmix(A,b2,c)

print("Solution:", solution2)
print(f"Objective: {objective2}, increase: {((objective2 - objective) / objective) * 100: 6.2f}%")

Starting solution of the Linear programming problem using Primal Simplex

Coin0506I Presolve 5 (0) rows, 4 (0) columns and 20 (0) elements
Clp1000I sum of infeasibilities 5.68434e-14 - average 1.13687e-14, 1 fixed columns
Coin0506I Presolve 5 (0) rows, 3 (-1) columns and 15 (-5) elements
Clp0006I 0  Obj 352890.14 Dual inf 56036.193 (3)
Clp0029I End of values pass after 3 iterations
Clp0000I Optimal - objective value 352890.14
Clp0000I Optimal - objective value 352890.14
Coin0511I After Postsolve, objective 352890.14, infeasibilities - dual 0 (0), primal 0 (0)
Clp0006I 0  Obj 352890.14
Clp0000I Optimal - objective value 352890.14
Solution: [630.4265041888804, 98.71794871794896, 0.0, 355.3566895151053]
Objective: 352890.1370906322, increase:  10.00%
Clp0000I Optimal - objective value 352890.14
Clp0000I Optimal - objective value 352890.14
Clp0032I Optimal objective 352890.1371 - 0 iterations time 0.002, Idiot 0.00


Task 3: __Find which constraint is the most restrictive by changing the input parameters $b_i$ one by one.__

We can loop over each ingredient and solve a problem where one of the components is increased by 0.1. This will tell us which constraint could be relaxed to increase the total revenue, and possibly which one yields the greatest increase.

In [10]:
for i in I:
    b3 = b[:]    # (TODO) Use [:] to make a copy of b, as otherwise modifying b3 would change b as well
    b3[i] =1.1*b3[i] # (TODO) Make a tiny change to the right-hand side
    # (TODO) solve problem with new right-hand side vector
    _, solution3, obj = solve_productmix(A,b3,c)

    print(f"Objective: {obj:10.2f}, Increase: {obj - objective:8.2f}, {(obj - objective)/objective * 100:5.2f}%")

Starting solution of the Linear programming relaxation problem using Primal Simplex

Coin0506I Presolve 5 (0) rows, 4 (0) columns and 20 (0) elements
Clp1000I sum of infeasibilities 0 - average 0, 1 fixed columns
Coin0506I Presolve 5 (0) rows, 3 (-1) columns and 15 (-5) elements
Clp0006I 0  Obj 320809.22 Dual inf 56036.193 (3)
Clp0029I End of values pass after 3 iterations
Clp0000I Optimal - objective value 320809.22
Clp0000I Optimal - objective value 320809.22
Coin0511I After Postsolve, objective 320809.22, infeasibilities - dual 0 (0), primal 0 (0)
Clp0006I 0  Obj 320809.22
Clp0000I Optimal - objective value 320809.22
Clp0000I Optimal - objective value 320809.22
Clp0000I Optimal - objective value 320809.22
Clp0032I Optimal objective 320809.2155 - 0 iterations time 0.002, Idiot 0.00

Starting MIP optimization
Cgl0004I processed model has 5 rows, 4 columns (4 integer (0 of which binary)) and 20 elements
Coin3009W Conflict graph built in 0.000 seconds, density: 0.000%
Cgl0015I Clique St

However, there is no need to solve as many LPs as the constraints. This information is returned back to us with the __dual variables__ of the original problem. These give the unitary change in objective function due to a change in the right-hand side in each constraint.

In [3]:
m, sol, obj = solve_productmix(A,b,c)
# Check the resulting list with the values you obtained before.
print("Dual values:", [con.pi for con in m.constrs])

Dual values:Starting solution of the Linear programming problem using Primal Simplex

Coin0506I Presolve 5 (0) rows, 4 (0) columns and 20 (0) elements
Clp1000I sum of infeasibilities 0 - average 0, 1 fixed columns
Coin0506I Presolve 5 (0) rows, 3 (-1) columns and 15 (-5) elements
Clp0006I 0  Obj 320809.22 Dual inf 56036.193 (3)
Clp0029I End of values pass after 3 iterations
Clp0000I Optimal - objective value 320809.22
Clp0000I Optimal - objective value 320809.22
Coin0511I After Postsolve, objective 320809.22, infeasibilities - dual 0 (0), primal 0 (0)
Clp0006I 0  Obj 320809.22
Clp0000I Optimal - objective value 320809.22
Clp0000I Optimal - objective value 320809.22
Clp0000I Optimal - objective value 320809.22
Clp0032I Optimal objective 320809.2155 - 0 iterations time 0.002, Idiot 0.00
 [-0.0, 191.92688499619217, 2117.479055597866, 118.81188118811873, -0.0]


Task 4: __Are all perfumes produced? If any of them is not produced, can we increase its retail price so that it is worth to produce it? By how much?__

In the original solution, perfume 3 is not produced. Its retail price $c_3$, which is also its objective function coefficient, is 260. Let's try running a loop where we increase $c_3$ by 5 until $x_3>0$ in the optimal solution.

In [4]:
c_new = c[:]  # Create a new list as a copy of c to use as objective function value below
sol = solution[:]  # Copy sol as well

while sol[2] <= 1e-5:  # Sol[2] is x_3 (remember that Python has 0-based indexing)
    c_new[2] += 5  # (TODO) increase by 5 objective function coefficient of item 3 (index 2) 
    m2, sol, obj2 = solve_productmix(A,b,c_new)  # (TODO) solve problem with new objective function
    print(f"With c_3 = {c_new[2]}: produce {sol[2]} liters of 3")

print (f"We need to set the price c_3 to {c_new[2]}")

With c_3 = 265: produce 0.0 liters of 3
With c_3 = 270: produce 0.0 liters of 3
Starting solution of the Linear programming problem using Primal Simplex

Coin0506I Presolve 5 (0) rows, 4 (0) columns and 20 (0) elements
Clp1000I sum of infeasibilities 0 - average 0, 1 fixed columns
Coin0506I Presolve 5 (0) rows, 3 (-1) columns and 15 (-5) elements
Clp0006I 0  Obj 320809.22 Dual inf 56036.193 (3)
Clp0029I End of values pass after 3 iterations
Clp0000I Optimal - objective value 320809.22
Clp0000I Optimal - objective value 320809.22
Coin0511I After Postsolve, objective 320809.22, infeasibilities - dual 0 (0), primal 0 (0)
Clp0006I 0  Obj 320809.22
Clp0000I Optimal - objective value 320809.22
Clp0000I Optimal - objective value 320809.22
Clp0000I Optimal - objective value 320809.22
Clp0032I Optimal objective 320809.2155 - 0 iterations time 0.002, Idiot 0.00
With c_3 = 275: produce 0.0 liters of 3
Starting solution of the Linear programming problem using Primal Simplex

Coin0506I Presolve 5 (

Again, we don't need to solve that many problems. All we need is the __reduced cost__ of $x_3$. In Python-MIP, the reduced cost of a variable `x` is obtained as `x.rc`. Given that the third variable of our model is `m.vars[2]`,we need to print:

In [15]:
print(m.vars[2].rc)

None


Notice anything? If we add 97.6466 to the original retail price $c_3$, we get $260 + 97.6466 = 357.6466$, which is between 355 and 360. Let's make a fine-grained comparison between two very close values of $c_3$: one at $260 + 97.64 = 357.64$ and another one at $260 + 97.65 = 357.65$. The prediction is that the first will give a value of zero for $x_3$, unlike the second.

In [16]:
c_test = c[:]

c_test[2] = c[2] + 97.64
_, sol, obj = solve_productmix(A,b,c_test)
print(f"before threshold: x_3 = {sol[2]}")

c_test[2] = c[2] + 97.65
_, sol, obj = solve_productmix(A,b,c_test)
print(f"after threshold: x_3 = {sol[2]}")

Starting solution of the Linear programming relaxation problem using Primal Simplex

Coin0506I Presolve 5 (0) rows, 4 (0) columns and 20 (0) elements
Clp1000I sum of infeasibilities 6.11618e-09 - average 1.22324e-09, 0 fixed columns
Coin0506I Presolve 5 (0) rows, 4 (0) columns and 20 (0) elements
Clp0006I 0  Obj 320809.22 Dual inf 68731.71 (4)
Clp0029I End of values pass after 4 iterations
Clp0000I Optimal - objective value 320809.22
Clp0000I Optimal - objective value 320809.22
Clp0000I Optimal - objective value 320809.22
Clp0032I Optimal objective 320809.2155 - 0 iterations time 0.002, Idiot 0.00

Starting MIP optimization
Cgl0004I processed model has 5 rows, 4 columns (4 integer (0 of which binary)) and 20 elements
Coin3009W Conflict graph built in 0.000 seconds, density: 0.000%
Cgl0015I Clique Strengthening extended 0 cliques, 0 were dominated
Cbc0012I Integer solution of -320565 found by DiveCoefficient after 0 iterations and 0 nodes (0.01 seconds)
Cbc0038I Full problem 5 rows 4 co

In [17]:
a = [[1,2,6],
     [4,5,3],
     [7,1,2]]
print(a[2][1])

1
